### PR TITLE
Update attrs

### PR DESCRIPTION
**Pull request by danodonovan**
Update `attrs` for compatibility across projects. 

* Update `attrs`
* Replace depreacted `attrs.assoc` with `attrs.evolve`
* Replace `__init__` with method from `attrs` created class
  - this avoids a recursion error

### CODE REVIEW

**Comment by danodonovan** [comment]
<strike>I'm not able to test the code in `tests/declarative/test_sqlalchemy.py` - Running with `TEST_POSTGRESQL_URL` set to either postgres 9.6 or 10 results in a `Illegal Instruction: 4` message and the `pytest` process exiting.</strike>

This seems to be a MacOS issue - the branch tests are running fine on other machines.

**Comment by danodonovan** [comment]
Address' #3

**Comment by danodonovan** [comment]
These changes do pass the repository's tests, but other dependent packages fail. I believe this is to do with `InputObjectTypeMeta` not having the correct `__hash__` or other dunder method (which should be tested).

File: graphjoiner/declarative/__init__.py [hunk 1]
```diff
@@ -412,8 +413,26 @@ def __new__(meta, name, bases, attrs):
 
         field_definitions, fields = _declare_fields(cls)
         fields = lazy(fields)
-        cls.__fields__ = staticmethod(lambda: list(field_definitions))
 
+        attrs_cls_attrs = dict(
+            raw_=attr.attrib(default=undefined),
+            **dict(
+                (field.attr_name, attr.attrib(default=getattr(field, "default", undefined)))
+                for field in fields().values()
```
Comment (mwilliamson-healx) [inline graphjoiner/declarative/__init__.py]: `fields` may refer to a type that hasn't been defined yet, which is why `fields` is defined using `lazy`. In other words, you can't evaluate `fields` within `__new__`.

File: graphjoiner/declarative/__init__.py [hunk 2]
```diff
@@ -23,6 +23,7 @@ def executor(root, mutation=None):
     return graphjoiner.executor(root_type, mutation=mutation_type)
 
 
+@attr.attrs
```
Comment (mwilliamson-healx) [inline graphjoiner/declarative/__init__.py]: `Type` isn't intended to be used as a value object, so I think we should get rid of the decorator.

**Pull request closed without merging**